File: arc/cpp/runtime/node/inputs_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <cstdint>
#include <string>

#include "gtest/gtest.h"

#include "x/cpp/telem/series.h"
#include "x/cpp/telem/telem.h"
#include "x/cpp/test/test.h"

#include "arc/cpp/ir/ir.h"
#include "arc/cpp/runtime/errors/errors.h"
#include "arc/cpp/runtime/node/inputs.h"
#include "arc/cpp/runtime/state/state.h"

namespace arc::runtime::node {

namespace {
arc::types::Param param(const std::string &name, arc::types::Kind kind) {
    arc::types::Param p;
    p.name = name;
    p.type = arc::types::Type{.kind = kind};
    return p;
}

arc::types::Param static_param(
    const std::string &name,
    arc::types::Kind kind,
    const x::json::json &value
) {
    auto p = param(name, kind);
    p.value = value;
    return p;
}

// builds a producer -> consumer state, returning both nodes. The producer's single
// output feeds the consumer input named `edge_target`.
arc::runtime::state::State build_state(
    const arc::ir::Node &consumer,
    const std::string &edge_target,
    arc::types::Kind producer_output_kind
) {
    arc::ir::Node producer;
    producer.key = "src";
    producer.type = "src";
    producer.outputs.push_back(param("output", producer_output_kind));

    arc::ir::IR ir;
    ir.nodes.push_back(producer);
    ir.nodes.push_back(consumer);
    if (!edge_target.empty())
        ir.edges.emplace_back(
            arc::ir::Handle("src", "output"),
            arc::ir::Handle(consumer.key, edge_target)
        );
    arc::runtime::state::Config cfg{.ir = ir, .channels = {}};
    return arc::runtime::state::State(cfg, arc::runtime::errors::noop_handler);
}
}

TEST(ResolvedInputsTest, OverlaysEdgeFedNumericInput) {
    arc::ir::Node consumer;
    consumer.key = "tgt";
    consumer.type = "tgt";
    consumer.inputs.push_back(
        static_param("static_p", arc::types::Kind::I64, static_cast<int64_t>(7))
    );
    consumer.inputs.push_back(param("edge_p", arc::types::Kind::I64));

    auto s = build_state(consumer, "edge_p", arc::types::Kind::I64);
    auto src = ASSERT_NIL_P(s.node("src"));
    auto &o = src.output(0);
    o->resize(1);
    o->set(0, static_cast<int64_t>(42));
    auto &ot = src.output_time(0);
    ot->resize(1);
    ot->set(0, x::telem::TimeStamp(1 * x::telem::MICROSECOND));

    auto tgt = ASSERT_NIL_P(s.node("tgt"));
    ASSERT_TRUE(tgt.refresh_inputs());

    const auto ri = ResolvedInputs::resolve(consumer);
    EXPECT_TRUE(ri.has_edges());
    EXPECT_EQ(std::get<int64_t>(*ri.value_of(tgt, "static_p")), 7);
    EXPECT_EQ(std::get<int64_t>(*ri.value_of(tgt, "edge_p")), 42);
}

TEST(ResolvedInputsTest, ReadsEdgeFedStringInput) {
    arc::ir::Node consumer;
    consumer.key = "tgt";
    consumer.type = "tgt";
    consumer.inputs.push_back(param("name", arc::types::Kind::String));

    auto s = build_state(consumer, "name", arc::types::Kind::String);
    auto src = ASSERT_NIL_P(s.node("src"));
    *src.output(0) = x::telem::Series(std::string("My_range_3"));
    *src.output_time(0) = x::telem::Series(
        x::telem::TimeStamp(1 * x::telem::MICROSECOND)
    );

    auto tgt = ASSERT_NIL_P(s.node("tgt"));
    ASSERT_TRUE(tgt.refresh_inputs());

    const auto ri = ResolvedInputs::resolve(consumer);
    EXPECT_EQ(ri.string_of(tgt, "name"), "My_range_3");
}

TEST(ResolvedInputsTest, NoEdgesWhenEveryInputIsStatic) {
    arc::ir::Node consumer;
    consumer.key = "tgt";
    consumer.type = "tgt";
    consumer.inputs.push_back(
        static_param("p", arc::types::Kind::I64, static_cast<int64_t>(1))
    );

    auto s = build_state(consumer, "", arc::types::Kind::I64);
    auto tgt = ASSERT_NIL_P(s.node("tgt"));

    const auto ri = ResolvedInputs::resolve(consumer);
    EXPECT_FALSE(ri.has_edges());
    EXPECT_EQ(std::get<int64_t>(*ri.value_of(tgt, "p")), 1);
}

TEST(ResolvedInputsTest, ChannelTypedInputIsNotEdgeFed) {
    arc::ir::Node consumer;
    consumer.key = "tgt";
    consumer.type = "tgt";
    consumer.inputs.push_back(param("ch", arc::types::Kind::Chan));
    consumer.inputs.push_back(
        static_param("static_p", arc::types::Kind::I64, static_cast<int64_t>(5))
    );

    const auto ri = ResolvedInputs::resolve(consumer);
    EXPECT_FALSE(ri.has_edges());
}

TEST(ResolvedInputsTest, ValueOfReturnsNulloptForUnknownInput) {
    arc::ir::Node consumer;
    consumer.key = "tgt";
    consumer.type = "tgt";
    consumer.inputs.push_back(
        static_param("p", arc::types::Kind::I64, static_cast<int64_t>(1))
    );

    auto s = build_state(consumer, "", arc::types::Kind::I64);
    auto tgt = ASSERT_NIL_P(s.node("tgt"));

    const auto ri = ResolvedInputs::resolve(consumer);
    EXPECT_FALSE(ri.value_of(tgt, "missing").has_value());
}
}

File: arc/go/runtime/node/inputs_test.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package node_test

import (
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/synnaxlabs/arc/ir"
	"github.com/synnaxlabs/arc/runtime/node"
	"github.com/synnaxlabs/arc/types"
	"github.com/synnaxlabs/x/telem"
	. "github.com/synnaxlabs/x/testutil"
)

var _ = Describe("ResolvedInputs", func() {
	It("overlays an edge-fed numeric input onto the static value map", func() {
		prog := ir.IR{
			Nodes: ir.Nodes{
				{Key: "src", Outputs: types.Params{{Name: ir.DefaultOutputParam, Type: types.I64()}}},
				{Key: "tgt", Inputs: types.Params{
					{Name: "static_p", Type: types.I64(), Value: int64(7)},
					{Name: "edge_p", Type: types.I64()},
				}},
			},
			Edges: ir.Edges{{
				Source: ir.Handle{Node: "src", Param: ir.DefaultOutputParam},
				Target: ir.Handle{Node: "tgt", Param: "edge_p"},
			}},
		}
		s := node.New(prog)
		src, tgt := s.Node("src"), s.Node("tgt")
		*src.Output(0) = telem.NewSeriesV[int64](42)
		*src.OutputTime(0) = telem.NewSeriesSecondsTSV(1)
		tgt.RefreshInputs()

		ri := MustSucceed(node.ResolveInputs(tgt, prog.Nodes.Get("tgt")))
		Expect(ri.HasEdges()).To(BeTrue())
		m := ri.ValueMap(tgt)
		Expect(m["static_p"]).To(Equal(int64(7)))
		Expect(m["edge_p"]).To(Equal(int64(42)))
	})

	It("reads an edge-fed string input", func() {
		prog := ir.IR{
			Nodes: ir.Nodes{
				{Key: "src", Outputs: types.Params{{Name: ir.DefaultOutputParam, Type: types.String()}}},
				{Key: "tgt", Inputs: types.Params{{Name: "name", Type: types.String()}}},
			},
			Edges: ir.Edges{{
				Source: ir.Handle{Node: "src", Param: ir.DefaultOutputParam},
				Target: ir.Handle{Node: "tgt", Param: "name"},
			}},
		}
		s := node.New(prog)
		src, tgt := s.Node("src"), s.Node("tgt")
		*src.Output(0) = telem.NewSeriesV[string]("My_range_3")
		*src.OutputTime(0) = telem.NewSeriesSecondsTSV(1)
		tgt.RefreshInputs()

		ri := MustSucceed(node.ResolveInputs(tgt, prog.Nodes.Get("tgt")))
		Expect(ri.ValueMap(tgt)["name"]).To(Equal("My_range_3"))
	})

	It("reports no edges when every input is static", func() {
		prog := ir.IR{Nodes: ir.Nodes{
			{Key: "tgt", Inputs: types.Params{{Name: "p", Type: types.I64(), Value: int64(1)}}},
		}}
		s := node.New(prog)
		ri := MustSucceed(node.ResolveInputs(s.Node("tgt"), prog.Nodes.Get("tgt")))
		Expect(ri.HasEdges()).To(BeFalse())
		Expect(ri.ValueMap(s.Node("tgt"))["p"]).To(Equal(int64(1)))
	})

	It("does not treat a channel-typed input as edge-fed", func() {
		prog := ir.IR{Nodes: ir.Nodes{
			{Key: "tgt", Inputs: types.Params{
				{Name: "ch", Type: types.Chan(types.I64())},
				{Name: "static_p", Type: types.I64(), Value: int64(5)},
			}},
		}}
		s := node.New(prog)
		ri := MustSucceed(node.ResolveInputs(s.Node("tgt"), prog.Nodes.Get("tgt")))
		Expect(ri.HasEdges()).To(BeFalse())
	})

	It("returns ErrInputNotFound when an edge-fed param is absent from state", func() {
		prog := ir.IR{Nodes: ir.Nodes{
			{Key: "tgt", Inputs: types.Params{{Name: "x", Type: types.I64(), Value: int64(1)}}},
		}}
		s := node.New(prog)
		// A node carrying an edge-fed input the state was not built with.
		mismatched := ir.Node{Key: "tgt", Inputs: types.Params{{Name: "phantom", Type: types.I64()}}}
		Expect(node.ResolveInputs(s.Node("tgt"), mismatched)).
			Error().To(MatchError(node.ErrInputNotFound))
	})

	Describe("ValidationMap", func() {
		It("substitutes a typed zero for an edge-fed input and keeps static values", func() {
			prog := ir.IR{
				Nodes: ir.Nodes{
					{Key: "src", Outputs: types.Params{{Name: ir.DefaultOutputParam, Type: types.I64()}}},
					{Key: "tgt", Inputs: types.Params{
						{Name: "static_p", Type: types.String(), Value: "kept"},
						{Name: "edge_p", Type: types.I64()},
					}},
				},
				Edges: ir.Edges{{
					Source: ir.Handle{Node: "src", Param: ir.DefaultOutputParam},
					Target: ir.Handle{Node: "tgt", Param: "edge_p"},
				}},
			}
			s := node.New(prog)
			ri := MustSucceed(node.ResolveInputs(s.Node("tgt"), prog.Nodes.Get("tgt")))
			m := ri.ValidationMap()
			Expect(m["static_p"]).To(Equal("kept"))
			Expect(m["edge_p"]).To(Equal(int64(0)))
		})

		DescribeTable("yields the Go-typed zero matching the edge-fed param type",
			func(t types.Type, want any) {
				prog := ir.IR{
					Nodes: ir.Nodes{
						{Key: "src", Outputs: types.Params{{Name: ir.DefaultOutputParam, Type: t}}},
						{Key: "tgt", Inputs: types.Params{{Name: "p", Type: t}}},
					},
					Edges: ir.Edges{{
						Source: ir.Handle{Node: "src", Param: ir.DefaultOutputParam},
						Target: ir.Handle{Node: "tgt", Param: "p"},
					}},
				}
				s := node.New(prog)
				ri := MustSucceed(node.ResolveInputs(s.Node("tgt"), prog.Nodes.Get("tgt")))
				Expect(ri.ValidationMap()["p"]).To(Equal(want))
			},
			Entry("string", types.String(), ""),
			Entry("f32", types.F32(), float32(0)),
			Entry("f64", types.F64(), float64(0)),
			Entry("u8", types.U8(), uint8(0)),
			Entry("u16", types.U16(), uint16(0)),
			Entry("u32", types.U32(), uint32(0)),
			Entry("u64", types.U64(), uint64(0)),
			Entry("i8", types.I8(), int8(0)),
			Entry("i16", types.I16(), int16(0)),
			Entry("i32", types.I32(), int32(0)),
			Entry("i64", types.I64(), int64(0)),
			Entry("timestamp falls through to i64", types.TimeStamp(), int64(0)),
		)
	})
})
